Test regressions in 1.2.0

When running tests from 1.2.0 source distribution, I'm seeing new test failures:

```pytb
$ uv pip install -e . geventhttpclient
Resolved 9 packages in 2.01s
      Built dulwich @ file:///tmp/dulwich-1.2.0
Prepared 3 packages in 28.77s
░░░░░░░░░░░░░░░░░░░░ [0/9] Installing wheels...                                                                                        Installed 9 packages in 131ms
 + brotli==1.2.0
 + certifi==2026.2.25
 + dulwich==1.2.0 (from file:///tmp/dulwich-1.2.0)
 + gevent==26.4.0
 + geventhttpclient==2.3.9
 + greenlet==3.4.0
 + urllib3==2.6.3
 + zope-event==6.1
 + zope-interface==8.3
$ python -m unittest -b
/tmp/dulwich-1.2.0/contrib/test_swift_smoke.py:45: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['urllib3.util.ssl_ (/tmp/dulwich-1.2.0/.venv/lib/python3.13/site-packages/urllib3/util/ssl_.py)', 'urllib3.util (/tmp/dulwich-1.2.0/.venv/lib/python3.13/site-packages/urllib3/util/__init__.py)']. 
  monkey.patch_all()
.sssssssssssssssss..............EE..................................s.......................................................................................................................Test template content.................................................................................................................................................s...sssssssssssssssssssssssssssssssssssssssssssssssssssssssss......................../usr/lib/python3.13/unittest/suite.py:107: ResourceWarning: unclosed <socket object, fd=5, family=2, type=1, proto=0>
  for index, test in enumerate(self):
ResourceWarning: Enable tracemalloc to get the object allocation traceback
./usr/lib/python3.13/unittest/suite.py:84: ResourceWarning: unclosed <socket object, fd=5, family=2, type=1, proto=0>
  return self.run(*args, **kwds)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
...................................s......s...............s.............................s.............................................s.............................................s.............s......................s......................................Auto-merging conflict.txt
CONFLICT (content): Merge conflict in conflict.txt
Automatic merge failed; fix conflicts and then commit the result.
.....counting objects: 6, done.
found 0 deltas to reuse
................F.....s................sssss............................................................................................ss.sss.ss..................................................................................s.....................................................counting objects: 3, done.
found 0 deltas to reuse
.........................counting objects: 4, done.
found 0 deltas to reuse
counting objects: 3, done.
found 0 deltas to reuse
..........ssssssss...............................................................................................................................................................................................................................................counting objects: 2, done.
Receiving objects: 100% (2/2), 0.00 MiB | 0.01 MiB/s, done.
copied 1 pack entries 1/2
.counting objects: 2, done.
found 0 deltas to reuse
Push to http://localhost:35641 successful.
Ref refs/heads/master updated
......................................................counting objects: 3, done.
found 0 deltas to reuse
.counting objects: 4, done.
found 0 deltas to reuse
counting objects: 3, done.
found 0 deltas to reuse
............ssssssss.............................ssssss....................ss...........counting objects: 4, done.
found 0 deltas to reuse
.counting objects: 3, done.
found 0 deltas to reuse
................counting objects: 4, done.
found 0 deltas to reuse
.............................s...s...ss......................counting objects: 3, done.
found 0 deltas to reuse
.counting objects: 3, done.
found 0 deltas to reuse
...ss....................................sssssss....................................................................................................................................................................................................................................................................................................................................................................................s.........................................................................................................................................................................................................sss........s.ssss.........................................................................................................sssssssssssss.s....................................................................................................................................................................................................................................................................s..............................................................................................................................................................................................................................................sssssssssssssssssssssss..sssss............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ssss.s....................................................................s.....................................................................................................................................................................................................................................................s......................................ss..........................................................................................sssssss........................ssssss.....s............................................................................................................................................x............................................................................................................................
======================================================================
ERROR: contrib.test_swift (unittest.loader._FailedTest.contrib.test_swift)
----------------------------------------------------------------------
ImportError: Failed to import test module: contrib.test_swift
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/loader.py", line 396, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.13/unittest/loader.py", line 339, in _get_module_from_name
    __import__(name)
    ~~~~~~~~~~^^^^^^
  File "/tmp/dulwich-1.2.0/contrib/test_swift.py", line 57, in <module>
    from contrib import swift
  File "/tmp/dulwich-1.2.0/contrib/swift.py", line 202, in <module>
    scon: "SwiftConnector", filename: str, object_format: ObjectFormat
                                                          ^^^^^^^^^^^^
NameError: name 'ObjectFormat' is not defined


======================================================================
ERROR: contrib.test_swift_smoke (unittest.loader._FailedTest.contrib.test_swift_smoke)
----------------------------------------------------------------------
ImportError: Failed to import test module: contrib.test_swift_smoke
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/loader.py", line 396, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.13/unittest/loader.py", line 339, in _get_module_from_name
    __import__(name)
    ~~~~~~~~~~^^^^^^
  File "/tmp/dulwich-1.2.0/contrib/test_swift_smoke.py", line 47, in <module>
    from contrib import swift  # noqa: E402
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/dulwich-1.2.0/contrib/swift.py", line 202, in <module>
    scon: "SwiftConnector", filename: str, object_format: ObjectFormat
                                                          ^^^^^^^^^^^^
NameError: name 'ObjectFormat' is not defined


======================================================================
FAIL: test_git_uses_dulwich_midx (tests.compat.test_midx.MIDXCompatTests.test_git_uses_dulwich_midx)
Test that Git can use a MIDX file created by Dulwich.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/dulwich-1.2.0/tests/compat/test_midx.py", line 135, in test_git_uses_dulwich_midx
    run_git_or_fail(["multi-pack-index", "verify"], cwd=work_dir)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/dulwich-1.2.0/tests/compat/utils.py", line 190, in run_git_or_fail
    raise AssertionError(
        f"git with args {args!r} failed with {returncode}: stdout={stdout!r} stderr={stderr!r}"
    )
AssertionError: git with args ['multi-pack-index', 'verify'] failed with 1: stdout=b'' stderr=b'failed to load pack in position 0\nfailed to load pack in position 1\nfailed to load pack in position 2\noid lookup out of order: oid[0] = 0681e6d81c236af71f2cf73672ce4f3e2923607d >= 0681e6d81c236af71f2cf73672ce4f3e2923607d = oid[1]\noid lookup out of order: oid[3] = 1b33a567898f7f1aba5981c8cf8a13800a8d4743 >= 1b33a567898f7f1aba5981c8cf8a13800a8d4743 = oid[4]\noid lookup out of order: oid[5] = 33b4e026b2670b979eece771fc7e622eaad2137c >= 33b4e026b2670b979eece771fc7e622eaad2137c = oid[6]\noid lookup out of order: oid[7] = 3a7ab7744cddae282ef540643b59623e0d3103bd >= 3a7ab7744cddae282ef540643b59623e0d3103bd = oid[8]\noid lookup out of order: oid[10] = 667497ca9ad4d7cb8450aea000b7670dd678d99e >= 667497ca9ad4d7cb8450aea000b7670dd678d99e = oid[11]\noid lookup out of order: oid[12] = 672478bd16a75e7d4d0c5656e939e15039b94565 >= 672478bd16a75e7d4d0c5656e939e15039b94565 = oid[13]\noid lookup out of order: oid[14] = 9d1a06f1eef26119255e7de7ceaf6552d94df2d5 >= 9d1a06f1eef26119255e7de7ceaf6552d94df2d5 = oid[15]\noid lookup out of order: oid[17] = a422559f4c0efd91a9e7bca212fdd65112f90aca >= a422559f4c0efd91a9e7bca212fdd65112f90aca = oid[18]\noid lookup out of order: oid[19] = a8f8e2553d392dfeee1f203f92bc28d3edccc01d >= a8f8e2553d392dfeee1f203f92bc28d3edccc01d = oid[20]\noid lookup out of order: oid[21] = b6d32270e011bb26e9e3087ffe306d1cd47512bd >= b6d32270e011bb26e9e3087ffe306d1cd47512bd = oid[22]\noid lookup out of order: oid[23] = cd72213d81c327e3cd736e0aff305d261cbf3c05 >= cd72213d81c327e3cd736e0aff305d261cbf3c05 = oid[24]\noid lookup out of order: oid[25] = d6d3dc397a4f7ff2ca5ff783ff680882595c8576 >= d6d3dc397a4f7ff2ca5ff783ff680882595c8576 = oid[26]\n'

----------------------------------------------------------------------
Ran 4133 tests in 235.140s

FAILED (failures=1, errors=2, skipped=210, expected failures=1)
```

Actually, looks like MIDX tests are very flaky, and randomly fail with different errors, e.g.:

```pytb
 python -m unittest -b tests.compat.test_midx
..F.E.
======================================================================
ERROR: test_read_git_midx (tests.compat.test_midx.MIDXCompatTests.test_read_git_midx)
Test that Dulwich can read a MIDX file created by Git.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/dulwich-1.2.0/tests/compat/test_midx.py", line 47, in tearDown
    rmtree_ro(self.test_dir)
    ~~~~~~~~~^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/shutil.py", line 763, in rmtree
    _rmtree_safe_fd(stack, onexc)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/shutil.py", line 707, in _rmtree_safe_fd
    onexc(func, path, err)
    ~~~~~^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/shutil.py", line 658, in _rmtree_safe_fd
    os.rmdir(name, dir_fd=dirfd)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 39] Directory not empty: '/tmp/tmp68m5utrx'

======================================================================
FAIL: test_midx_object_lookup_matches_git (tests.compat.test_midx.MIDXCompatTests.test_midx_object_lookup_matches_git)
Test that object lookups through MIDX match Git's results.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/dulwich-1.2.0/tests/compat/test_midx.py", line 146, in test_midx_object_lookup_matches_git
    run_git_or_fail(["multi-pack-index", "write"], cwd=work_dir)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/dulwich-1.2.0/tests/compat/utils.py", line 190, in run_git_or_fail
    raise AssertionError(
        f"git with args {args!r} failed with {returncode}: stdout={stdout!r} stderr={stderr!r}"
    )
AssertionError: git with args ['multi-pack-index', 'write'] failed with 128: stdout=b'' stderr=b"fatal: Unable to create '/tmp/tmps0f8zjbx/test-repo/objects/pack/multi-pack-index.lock': File exists.\n\nAnother git process seems to be running in this repository, or the lock file may be stale\n"

----------------------------------------------------------------------
Ran 6 tests in 0.898s

FAILED (failures=1, errors=1)
$ python -m unittest -b tests.compat.test_midx
..F.F.
======================================================================
FAIL: test_midx_object_lookup_matches_git (tests.compat.test_midx.MIDXCompatTests.test_midx_object_lookup_matches_git)
Test that object lookups through MIDX match Git's results.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/dulwich-1.2.0/tests/compat/test_midx.py", line 155, in test_midx_object_lookup_matches_git
    self.assertIsNotNone(midx, "MIDX should be loaded")
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: unexpectedly None : MIDX should be loaded

======================================================================
FAIL: test_read_git_midx (tests.compat.test_midx.MIDXCompatTests.test_read_git_midx)
Test that Dulwich can read a MIDX file created by Git.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/dulwich-1.2.0/tests/compat/test_midx.py", line 109, in test_read_git_midx
    self.assertTrue(pack_name.startswith("pack-"))
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: False is not true

----------------------------------------------------------------------
Ran 6 tests in 0.785s

FAILED (failures=2)
```

--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,11 @@
+1.2.1	UNRELEASED
+
+ * Deduplicate objects when writing a multi-pack-index. Objects present
+   in multiple packs (e.g. after ``git gc`` creates a cruft pack) would
+   otherwise produce an OIDL chunk with repeated SHAs, causing ``git
+   multi-pack-index verify`` to fail with "oid lookup out of order".
+   (Jelmer Vernooĳ, #2152)
+
 1.2.0	2026-04-21
 
  * Expand ``log`` command options: add ``--oneline``, ``--abbrev-commit``,

--- a/dulwich/midx.py
+++ b/dulwich/midx.py
@@ -499,14 +499,23 @@ def write_midx(
     # Sort pack entries by pack name (required by Git)
     pack_index_entries_sorted = sorted(pack_index_entries, key=lambda x: x[0])
 
-    # Collect all objects from all packs
-    all_objects: list[tuple[RawObjectID, int, int]] = []  # (sha, pack_id, offset)
+    # Collect all objects from all packs, deduplicating objects that appear
+    # in multiple packs. Git's MIDX format requires strictly increasing OIDs
+    # in the OIDL chunk, so duplicates must be resolved. When the same object
+    # is in multiple packs, we keep the occurrence from the first pack (in
+    # pack-name order), matching how Git itself handles duplicates.
+    seen: dict[RawObjectID, tuple[int, int]] = {}  # sha -> (pack_id, offset)
     pack_names: list[str] = []
 
     for pack_id, (pack_name, entries) in enumerate(pack_index_entries_sorted):
         pack_names.append(pack_name)
         for sha, offset, _crc32 in entries:
-            all_objects.append((sha, pack_id, offset))
+            if sha not in seen:
+                seen[sha] = (pack_id, offset)
+
+    all_objects: list[tuple[RawObjectID, int, int]] = [
+        (sha, pack_id, offset) for sha, (pack_id, offset) in seen.items()
+    ]
 
     # Sort all objects by SHA
     all_objects.sort(key=lambda x: x[0])

--- a/tests/test_midx.py
+++ b/tests/test_midx.py
@@ -178,6 +178,51 @@ class MIDXWriteTests(TestCase):
         self.assertIsNotNone(result)
         self.assertEqual(large_offset, result[1])
 
+    def test_write_duplicate_objects_across_packs(self):
+        """Objects present in multiple packs must only appear once in OIDL.
+
+        Git's MIDX OIDL chunk requires strictly increasing OIDs; duplicates
+        cause 'oid lookup out of order' errors in ``git multi-pack-index
+        verify``. This happens in real repositories when cruft packs preserve
+        unreachable objects that also exist in other packs.
+        """
+        f = BytesIO()
+
+        shared_sha = b"\x05" * 20
+        pack_entries = [
+            (
+                "pack-aaa.idx",
+                [
+                    (b"\x01" * 20, 100, 0),
+                    (shared_sha, 200, 0),
+                ],
+            ),
+            (
+                "pack-bbb.idx",
+                [
+                    (shared_sha, 999, 0),
+                    (b"\x09" * 20, 50, 0),
+                ],
+            ),
+        ]
+
+        write_midx(f, pack_entries, HASH_ALGORITHM_SHA1)
+        f.seek(0)
+        midx = MultiPackIndex("test.midx", file=f, contents=f.read())
+
+        self.assertEqual(3, len(midx))
+
+        # Shared object should resolve to the first pack (pack-aaa sorts first)
+        result = midx.object_offset(shared_sha)
+        self.assertIsNotNone(result)
+        self.assertEqual("pack-aaa.idx", result[0])
+        self.assertEqual(200, result[1])
+
+        # OIDs must be strictly increasing (no duplicates)
+        entries = list(midx.iterentries())
+        shas = [e[0] for e in entries]
+        self.assertEqual(shas, sorted(set(shas)))
+
     def test_write_midx_file(self):
         """Test writing a MIDX file to disk."""
         with tempfile.TemporaryDirectory() as tmpdir:

